### PR TITLE
Apple2: Rewrite opendir in assembly

### PR DESCRIPTION
**Pull request by colinleroy**
58 bytes size gain :)
tested nominal ("." and normal path) and all error cases (open failure, read failure, malloc failure)

### CODE REVIEW

**Comment by mrdudz** [comment]
thanks!

File: libsrc/apple2/opendir.s [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        ; malloc the dir struct
+        lda       #<.sizeof(DIR)
+        ldx       #>.sizeof(DIR)
+        jsr       _malloc
+        bne       :+
+
+        ; We failed to allocate
+        pla                   ; Get fd back
+        ldx       #$00
+        jsr       _close      ; close it
```
Comment (oliverschmidt) [inline libsrc/apple2/opendir.s:61]: Start comment in uppercase.

File: libsrc/apple2/opendir.s [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+        lda       #ENOMEM     ; Set error
+        jsr       ___directerrno
+
+@return_null:
+        lda       #$00
+        tax
+        rts
```
Comment (oliverschmidt) [inline libsrc/apple2/opendir.s:69]: Why no JMP return0 ?

File: libsrc/apple2/opendir.s (hunk 1 above)
Comment (colinleroy) [reply libsrc/apple2/opendir.s:61]: As this one is merged, addressed in #2535

File: libsrc/apple2/opendir.s [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        ; Save fd to dir struct
+        lda       #$00
+        ldy       #DIR::FD + 1
+        sta       (ptr1),y
+
+        dey
+        pla                   ; Get fd back
+        sta       (ptr1),y
+
+        jsr       pusha0      ; push fd for read
```
[the remark this answers is not in the document]
Comment (colinleroy) [reply libsrc/apple2/opendir.s:87]: As this one is merged, addressed in #2535

File: libsrc/apple2/opendir.s (hunk 2 above)
Comment (colinleroy) [reply libsrc/apple2/opendir.s:69]: Saw it later, it's done in #2535 :)

**Comment by oliverschmidt** [comment]
@mrdudz: So my POV isn't relevant anymore?

**Comment by mrdudz** [comment]
Mmh?

**Comment by mrdudz** [comment]
Sorry, i just stumbled about it in my weekly routine and merging it seemed straightforward

**Pull request merged**